clang/lib/CodeGen/CGHLSLBuiltins.cpp

```diff
@@ -385,12 +385,17 @@ Value *CodeGenFunction::EmitHLSLBuiltinExpr(unsigned BuiltinID,
            "Intrinsic dot2add is only allowed for dxil architecture");
     Value *A = EmitScalarExpr(E->getArg(0));
     Value *B = EmitScalarExpr(E->getArg(1));
-    Value *C = EmitScalarExpr(E->getArg(2));
+    Value *Acc = EmitScalarExpr(E->getArg(2));
+
+    Value *AX = Builder.CreateExtractElement(A, Builder.getSize(0));
```

Comment:
im not sure how i feel about scalarization in the frontend. This seems like a one off so have a minor concern that changes like this will make the vectorized DXIL coming in 6.9 have special cases in the frontend.  This is a concern for the future so take it with a grain of salt.

Reply:
I would argue that this really is a special case. `dot2add` can only possibly take two-element vectors (it's right in the name), so we don't need to worry about scalarizing vectors of different lengths. It's also unlikely that we'll be adding DXIL operations for anything other than this version that works with `half`. If this operation were something more generic I'd be a little more hesitant to do this logic early, but I don't think we need to worry here.